### PR TITLE
Protect operations on keys with mutexes if MBEDTLS_THREADING_C is enabled

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR introduces a global data mutex, as well as a mutex per each key slot.
Currently the global mutex aims to protect the global "key_slots_initialized" variable, and only it; 
The main weight of the job is on the key slot mutexes, as this simplifies many situations.
These mutexes should be changed to a readers-writer lock, to enable multiple threads reading from a slot mutex;
There are two things that are currently unhandled in this PR:

- get_and_lock_key_slot_in_memory, get_empty_key_slot, psa_wipe_all_key_slots, and mbedtls_psa_get_stats iterate through key slots and read from them. This means that, before reading from a given slot, they lock the respective key mutex. The problem is that if a given thread locks a key mutex (after, for example, starting key creation) and then the same thread calls any of the functions mentioned above - a double lock will happen. 
- psa_copy_key problem: it locks one key slot, and then calls `psa_start_key_creation()`, which in turn iterates over the key array. Since it's reading key attributes - if it hits the first, locked slot, a double lock will occur. A simple implementation of a RW-lock will not prevent this, so we might need more abstraction (tracking threads vs mutexes locked), or a different solution. Please note that when locking a non-volatile key via `psa_get_and_lock_key_slot_with_policy()`, the code also iterates over the key slots, so simply changing the order of functions (`psa_start_key_creation` and `psa_get_and_lock_key_slot_with_policy`) won't work. Current state of the PR contains a hotfix that has implemented such temporary hack, but one test still fails because of this reason: `Non reusable key slots integrity in case of key slot starvation`.
- Reading any slot data (or attributes) requires a slot lock to be locked. Thus, psa_is_key_slot_locked is obsolete with current design and assumption of a simple (not readers-writer) mutex:
    - The key slot might be either locked, so the caller has to wait for a mutex before calling psa_is_key_slot_locked. When the mutex is unlocked, the caller is free to lock it and find out that the key slot is not locked.
    - The key slot is not locked, so the caller locks the slot mutex and finds out that the key slot is not locked.
Please note that with a readers-writer lock this will no longer be a problem. It's only used by mbedtls_psa_get_stats() this way though.

Other considerations:
- Slot states, as suggested here: https://github.com/AndrzejKurek/mbedtls/blob/psa-thread_safety-doc/docs/architecture/psa-thread-safety.md - I'm not sure what the advantage of having certain slot states would be. If we decide to use a readers-writer lock, the implemented priority of locking will take care of acting as the state the slot is in. 
- Global vs local lock responsibilities
psa_get_empty_key_slot iterates over key slots and reads their attributes - it should lock respective slot mutexes before doing so. If not, and we consider the global mutex to be enough - there might be problems with other functions like psa_import_key: it calls psa_start_key_creation, so only the local lock is locked after that, and then psa_driver_wrapper_import_key which is slot agnostic. There's no way to lock the global lock for the entirety of the function, unless we add abstraction that will prevent double locking.
- Multipart operations and key destruction
After the introduction of key UID's this should not be a problem: at the beginning of each multipart operation the key slot mutex is locked and its UID is checked against the one stored in the operation object. If there's a mismatch, the operation is aborted.

- It was mentioned, for a global lock approach, that:
>Have a single mutex protecting all accesses to the key store and other global variables. In practice, this means every PSA API function needs to take the lock on entry and release on exit, except for:
> - Hash function.
> - Accessors for key attributes and other local structures.

I'm not sure why accessors for key attributes and hash functions would not require a mutex to be locked, unless we're talking about low level functions that already act under a mutex locked, and operating on local variables that were copied from a slot under a mutex.

Major todo:
- Change the simple mutex to a readers-writer lock.

Minor todos:
- Renaming key lock counter to "open" counter;
- Unify `psa_get_and_lock_key_slot_with_policy` usage - some calls to it do not unlock the key slot after a failure is returned, and some do. It would be best to add tests that exercise the failing calls.

### CODE REVIEW

**Comment by AndrzejKurek** [comment]
Merge passing, head failing with socket binding errors and timeouts in tests that do not enable `MBEDTLS_THREADING_C`, so I assume it's unrelated.

**Comment by AndrzejKurek** [comment]
This PR is now updated to reflect the idea of moving more authority over to the key slot locks, with the global lock only guarding the `key_slots_initialized` variable. This simplifies most of operations, but introduces some problems, now described in the PR description.

**Comment by AndrzejKurek** [comment]
This PR description should now (hopefully) contain all of the findings regarding the key-slot-focused approach that I took. 
There's a second PR in a more raw state that was aimed at handling everything with a global lock only: https://github.com/ARMmbed/mbedtls/pull/5190

**Comment by ronald-cron-arm** [comment]
It seems to me that the issues we have with double locks are due to the fact that a mutex is locked when accessing a key slot in RO (Read Only) and thus that multiple read accesses are not possible. The easiest solution to that problem is to not lock a mutex when accessing a key slot in RO and just increment a read counter (the read counters being protected from concurrent   RW accesses by `mbedtls_psa_slots_mutex`).

Then there is the case where while a key slot is read (its read counter is greater than 0) a write access is requested (through psa_destroy_key()). The write access needs to be blocked until the read accesses are completed and all new read accesses should be denied. An easy way to block the write access is to busy wait that the read counter goes down to 0 but we would need a `yield` primitive for this to be practical. No solution purely based on mutexes comes to my mind but the `yield` primitive is simple and well supported in multi-threaded OSs I think thus adding this platform dependency does not seem a blocker to me.

**Comment by gilles-peskine-arm** [comment]
> the yield primitive is simple and well supported in multi-threaded OSs

No, not really, it varies a lot. On many OSes, there is no yield, or it's useless. For example, on many OSes, a high-priority thread can't yield to a low-priority thread.

We basically need a reader-writer lock, and [it is possible to implement one on top of two mutexes](https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock#Using_two_mutexes).

**Comment by ronald-cron-arm** [comment]
> We basically need a reader-writer lock, and [it is possible to implement one on top of two mutexes](https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock#Using_two_mutexes).

I agree. Looking at the link I thought about that but thought as well that "This requires that a mutex acquired by one thread can be released by another." was a no go.

**Comment by gilles-peskine-arm** [comment]
> "This requires that a mutex acquired by one thread can be released by another." was a no go.

Good point, not all platforms allow that.

If we add an extra concurrency primitive requirement, I'd suggest condition variables, which work well as an addition to mutexes, and are fairly common. Alternatively, semaphores, which AFAIK are equivalent in power to mutexes + condition variables, and are commonly available as primitives, but a bit harder to use and not an extension of the current API.

**Comment by ronald-cron-arm** [comment]
Looking at https://www.keil.com/pack/doc/CMSIS/RTOS2/html/index.html I do not see any direct support for conditions. There is support for semaphores though.

File: library/psa_crypto.c [hunk 1]
```diff
@@ -2002,6 +2027,7 @@ psa_status_t psa_copy_key( mbedtls_svc_key_id_t source_key,
     psa_key_attributes_t actual_attributes = *specified_attributes;
     psa_se_drv_table_entry_t *driver = NULL;
     size_t storage_size = 0;
+    int unlock_slot = 1;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/psa_crypto.c:2030]: This is just a part of a temporary hack to make `psa_copy_key` work with the multi-mutexed approach. A preferred solution would be not using any data from the source key in the call to psa_start_key_creation.

**Comment by AndrzejKurek** [comment]
Closing in favor of https://github.com/Mbed-TLS/mbedtls/pull/5673.

**Pull request closed without merging**